Enable code coverage

heaths opened the pull request:
Resolves #14427 by limiting reports to a single service directory (CIs) or per test project (dev environments). For CIs, full coverage reports of everything build will be uploaded. For dev environments, a summary HTML file will be output to test projects' TestResults directories.

heaths committed:
Use VSTest integration for Coverlet

Fixes #15231. Relates to #14427.

heaths committed:
Do not ignore CompilerGeneratedAttribute

See https://github.com/coverlet-coverage/coverlet/issues/794#issuecomment-612067552

heaths committed:
Enable code coverage for PRs

heaths committed:
Publish code coverage even for failed tests

heaths committed:
Enable code coverage reports

Resolves #14427 by limiting reports to a single service directory (CIs) or per test project (dev environments). For CIs, full coverage reports of everything build will be uploaded. For dev environments, a summary HTML file will be output to test projects' TestResults directories.

heaths committed:
Fix filefilters path for CI

heaths committed:
Use absolute path for filefilters

Same as dev support; reportgenerator documentation is not clear on when absolute or relative paths are required.

heaths committed:
Use absolute path for reports

heaths committed:
Update report directory to limit globbing

heaths committed:
Disable code coverage of track 2 mgmt

Opened #17090 to track re-enabling once improved.

heaths committed:
Generate report before uploading test results

The exact same commands are working locally using the same versions of ReportGenerator, sans running the test publishing executable. There's also extra coverage files showing up that I'm wondering if it's responsible.

heaths committed:
Temporarily upload all code coverage artifacts

heaths committed:
Split between props and targets again

At one point, this mostly worked. Going back to how I was split between props and targets before to see if that makes a difference.

heaths committed:
Collect more information

pakrym commented on eng/CodeCoverage.runsettings:
```diff
@@ -0,0 +1,18 @@
+        <Configuration>
+          <Format>cobertura</Format>
+          <ExcludeByAttribute>ExcludeFromCodeCoverageAttribute,GeneratedCodeAttribute,Obsolete</ExcludeByAttribute>
+          <IncludeTestAssembly>false</IncludeTestAssembly>
```
Why do we have this set to false? I thought excluding the test assembly is desiriable.

heaths replied:
That's why it's false, i.e. don't *include* the test assembly. 😉

pakrym commented on eng/Directory.Build.Data.props:
```diff
@@ -86,7 +98,7 @@
     <IsPackable>false</IsPackable>
     <RequiredTargetFrameworks>netcoreapp2.1</RequiredTargetFrameworks>
-    <RequiredTargetFrameworks Condition="'$(OS)' == 'Windows_NT'">netcoreapp2.1;net461</RequiredTargetFrameworks>
+    <RequiredTargetFrameworks Condition="'$(OS)' == 'Windows_NT'">net461;netcoreapp2.1</RequiredTargetFrameworks>
```
Can you add a comment about why the order is important here.

heaths replied:
Done. I'll include it in the next updates, which hopefully is just taking out some extra (slow) logging.

pakrym commented:
The current set of changes LGTM. I'll keep an eye on the incoming changes.

heaths committed:
Always define CoverletGetPathMap

@clairernovotny recommended important some changes. These are working locally when simulating a CI, so 🤞.

heaths committed:
Resolve PR feedback

Fix extra (temporary) logging as well.

heaths commented:
Once done, it will show code coverage stats for the service directory (i.e. netcoreapp2.1 test run):
![image](https://user-images.githubusercontent.com/1532486/99862810-0703ce00-2b50-11eb-89bd-b571b3674289.png)

heaths committed:
Use different variable to detect CI

ContinuousIntegrationBuild wasn't defined for test projects.

heaths committed:
Replace curly braces with Of in file names

Fixes #17164

heaths committed:
Removing extra logging

heaths committed:
Renames files with curly braces

Fixes #17164

maririos commented on sdk/appconfiguration/Azure.Data.AppConfiguration/src/Azure.Data.AppConfiguration.csproj:
```diff
@@ -30,7 +30,7 @@
     <Compile Include="$(AzureCoreSharedSources)ConditionalRequestOptionsExtensions.cs" />
     <Compile Include="$(AzureCoreSharedSources)DiagnosticScope.cs" />
     <Compile Include="$(AzureCoreSharedSources)HashCodeBuilder.cs" />
-    <Compile Include="$(AzureCoreSharedSources)NoBodyResponse{T}.cs" />
```
where these names causing issues? or why the rename?

pakrym replied:
Files that have `{` or `}` in the name cause a crash in the coverage library we use. We filed an issue but until it's fixed and released we have to rename our source files.

jsquire replied:
Thank you for the context; that addresses my question below as well.

maririos replied:
Is there a place where this can be documented so until the bug is fixed, other people in the repo can avoid it?

heaths replied:
@jsquire keep in mind as well the class name doesn't change. Eliding `{T}` from a file name is pretty common. SA1649 accepts with and without. We have many generic classes without them currently.

maririos commented:
Is there any guidance or doc that talks about what the accepted  code coverage `x%` should be for a library? or what we are aiming for?

jsquire commented:
I'd prefer not to rename the `EventProcessor<TPartition>` file for Event Hubs, as it could easily be confused with the `EventProcessorClient` since both are often referred to as just "the event processor."   The current name is following a standard convention for disambiguating generic types in .NET.   Can you help me understand why the change is being proposed?

pakrym commented:
It there a way we can exclude shared code files like `Azure.Core.Argument` from the coverage for non-core projects?

jsquire commented:
> I'd prefer not to rename the `EventProcessor<TPartition>` file for Event Hubs, as it could easily be confused with the `EventProcessorClient` since both are often referred to colloquially as "the event processor." The current name is following a standard convention for disambiguating generic types in .NET. Can you help me understand why the change is being proposed?

Pavel's response answers this; while it's unfortunate, the selected rename seems like the best choice given the constraints.

jsquire commented:
> Once done, it will show code coverage stats for the service directory (i.e. netcoreapp2.1 test run):

Any chance that we can break this down by pipeline or individual library?   My concern is trying to remove the stats for T1 and management libraries which we don't own.   If we're already scoped to pipeline, than the forthcoming work to split out the T1 libraries into their own pipeline will take care of this.

jsquire reviewed (approved):
Event Hubs bits look good, given the constraint on filenames.

heaths commented:
@jsquire T1 is not covered - only T2. You can get individual package coverage in 1 of 2 ways:

1. From browsing to the assembly in the PR report from the Code Coverage tab.
2. From running `dotnet test /p:CollectCoverage=true` locally for your project or solution. In that case, you get 1 report per assembly - still only T2. That is covered in the CONTRIBUTING.md file changes so I encourage everyone to read that and let me know if it needs clarification.

@maririos 70% or better is what @AlexGhiondea would like. Eventually, I will enable build breaks for anything less and file exceptions, but this PR alone took *way* longer than expected due to lots of significant problems in third-party and even first-party (often causing issues with third-party) packages. This PR is only focused on reporting. I'll be working on support for exceptions and build breaks (with temporary exceptions)  subsequently.

@pakrym et. al. you can use the `[ExcludeFromCodeCoverage]` attribute on a method or class to exclude it. `[Obsolete]` is also ignored by default. See *eng/CodeCoverage.runsettings* for details.

jsquire commented:
> @jsquire T1 is not covered - only T2. You can get individual package coverage in 1 of 2 ways:

So long as the T1 libraries hanging out in the service directory don't throw off results, I'm good.   Thanks for the clarification.

heaths commented:
@jsquire take a look at *eng/CodeCoverage.targets* right at the top for what's coverable by default. As long as your T1s don't purport to be T2s you'll be fine. That said, looking at your reports, something is wrong: https://dev.azure.com/azure-sdk/public/_build/results?buildId=632467&view=codecoverage-tab. I see T1 in there but a cursory glance hasn't yielded why. Storage, Search, and Key Vault are only covering T2s. I don't see any Directory.Build.* files that would explain why either. Thoughts?

heaths committed:
Resolve PR feedback

heaths committed:
Do not reformat HTML coverage report

pakrym commented on sdk/eventhub/Microsoft.Azure.EventHubs/Directory.Build.props:
```diff
@@ -1,5 +1,6 @@
 ﻿<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
+    <ExcludeFromCodeCoverage>true</ExcludeFromCodeCoverage>
```
I'm not a fan of starting to exclude projects one by one. EH is not the only directory that has T1&T2 clients so we need a general solution here before we jump into workarounds.

heaths replied:
That only affects T1. Somehow - no way I could find after going through their various targets, nor does @jsquire know - they are detected as T2. None of the other CIs that ran as part of this PR exhibited that.

heaths replied:
Note from the targets file I added only T2 projects are covered by default. These are an anomaly.
